smart-py/objkt_swap_v3_test.py

```diff
@@ -531,3 +541,661 @@ def test_pause_contract():
+        metadata=sp.pack("ipfs://fff"),
+        royalties=100
+    ).run(
+        sender=artist1
+    )
+
+    # Artist updates the objkt to be managed by marketplace v3
+    scenario += objkt.update_operators(
+        [sp.variant("add_operator", objkt.operator_param.make(
+            owner=artist1.address,
+            operator=marketplaceV3.address,
+            token_id=152))]).run(sender=artist1)
+
+    # Successfully swap
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(1),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=artist1,
+        valid=True
+    )
+
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try to collect nonexistent swap and fail
+    scenario += marketplaceV3.collect(100).run(
+        sender = collector1,
+        amount = sp.mutez(1),
+        valid = False
+    )
+
+    # first one still exists
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # try to collect own swap and fail
+    #
+    # TODO should not be able to collect own swap
+    #
+    # scenario += marketplaceV3.collect(0).run(
+    #     sender = artist1,
+    #     amount = sp.mutez(1),
+    #     valid = False
+    # )
+    #
+    # # first one still exists
+    # scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    # scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    # scenario.verify(marketplaceV3.data.counter == 1)
+
+    # this should fail because amount is wrong
+    scenario += marketplaceV3.collect(0).run(
+        sender = collector1,
+        amount = sp.mutez(100),
+        valid = False
+    )
+
+    # first one still exists
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # this should pass
+    scenario += marketplaceV3.collect(0).run(
+        sender = collector1,
+        amount = sp.mutez(1),
+        valid = True
+    )
+
+    # swap should be gone now
+    #
+    # TODO the swap is still present even though it was successfully collected
+    # scenario.verify(marketplaceV3.data.counter == 1)
+    # scenario.verify(marketplaceV3.data.swaps.contains(0) == False)
+    # scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+
+    # this should fail now
+    scenario += marketplaceV3.collect(0).run(
+        sender = collector1,
+        amount = sp.mutez(1),
+        valid = False
+    )
+
+    # TODO
+    #
+    # The original functionality removed swaps from the array if the amount
+    # of items available becomes 0
```

Comment:
@jagracar this was the behaviour in v1, just want to confirm we leave the entry in the array

my personal pref is to leave it in the array with the number available set to 0 but maybe there's a reason to remove it completely

Reply:
I discussed it already with some people. At the moment when all the swapped editions are collected, the swap editions are 0 and one cannot collect more editions. The row for that swap remain in the swaps big map.
The situation is different for cancel swaps. If one cancels a swap, the row is removed from the big map.
I find that i bit ugly. Either one just set the editions to 0 in both cases, or one removes the row in both cases.
However, after discussing it with other people, we decided to keep it as it is in v2, because changing the behaviour might have unknown consequences for indexers or tools that the community developed.

Reply:
I updated the tests for this